File: Perl/092 Square digit chains.pl
#!/usr/bin/perl

# Daniel "Trizen" Șuteu
# License: GPLv3
# Date: 30 January 2017
# https://github.com/trizen

# https://projecteuler.net/problem=92

# Runtime: 8.466s

use 5.010;
use strict;

use List::Util qw(sum);
use ntheory qw(todigits);

my @cache;
my @squares = map { $_ * $_ } 0 .. 9;

sub chain {
    my ($n) = @_;

    $n == 1  && return 1;
    $n == 89 && return 89;

    $cache[$n] //= chain(sum(@squares[todigits($n)]));
}

my $count = 0;

foreach my $n (1 .. 1e7) {
    ++$count if chain($n) == 89;
}

say $count;

File: Julia/092 Square digit chains.jl
#!/usr/bin/julia

# Daniel "Trizen" Șuteu
# License: GPLv3
# Date: 30 January 2017
# https://github.com/trizen

# https://projecteuler.net/problem=92

# Runtime: 2.405s

const cache = zeros(Int64, 10^7)

cache[ 1] = 1
cache[89] = 89

function chain(n::Int64)
    if (cache[n] == 0)
        cache[n] = chain(sum(map((d)->d^2, digits(n))))
    else
        cache[n]
    end
end

function count()
    c = 0
    for n in 1:10^7
        if (chain(n) == 89)
            c += 1
        end
    end
    return c
end

println(count())
